s3/src/main/scala/monix/connect/s3/S3.scala

```diff
@@ -52,63 +58,201 @@ import software.amazon.awssdk.services.s3.model.{
+  * val bucket = "my-bucket"
+  * val key = "my-key"
+  * val content = "my-content"
   *
-  *   val httpClient = NettyNioAsyncHttpClient.builder()
-  *     .maxConcurrency(500)
-  *     .maxPendingConnectionAcquires(50000)
-  *     .connectionAcquisitionTimeout(Duration.ofSeconds(60))
-  *     .readTimeout(Duration.ofSeconds(60))
-  *     .build()
+  * def runS3App(s3: S3): Task[Array[Byte]] = {
+  *   for {
+  *     _ <- s3.createBucket(bucket)
+  *     _ <- s3.upload(bucket, key, content.getBytes)
+  *     existsObject <- s3.existsObject(bucket, key)
+  *     download <- {
+  *       if(existsObject) s3.download(bucket, key)
+  *       else Task.raiseError(NoSuchKeyException.builder().build())
+  *     }
+  *   } yield download
+  * }
   *
-  *   // necessary for all the exposed methods in the [[S3]] object
-  *   implicit val s3AsyncClient: S3AsyncClient = S3AsyncClient
-  *     .builder()
-  *     .httpClient(httpClient)
-  *     .credentialsProvider(DefaultCredentialsProvider.create())
-  *     .region(AWS_GLOBAL)
-  *     .build()
+  * val t = S3.create().use(s3 => runS3App(s3))
   * }}}
   *
   */
-object S3 {
+object S3 { self =>
 
   /**
-    * Creates a bucket.
+    * Creates a [[Resource]] that will use the values from a
+    * configuration file to allocate and release a [[S3]].
+    * Thus, the api expects an `application.conf` file to be present
+    * in the `resources` folder.
     *
-    * @see https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/model/CreateBucketRequest.Builder.html
-    * @param bucket the name of the bucket to be created.
-    * @param acl the canned ACL (Access Control List) to apply to the object.
-    *            if the service returns an enum value that is not available in the current SDK version, acl will return ObjectCannedACL.UNKNOWN_TO_SDK_VERSION.
-    * @param grantFullControl gives the grantee READ, READ_ACP, and WRITE_ACP permissions on the object.
-    * @param grantRead allows grantee to read the object data and its metadata.
-    * @param grantReadACP allows grantee to read the object ACL.
-    * @param grantWriteACP allows grantee to write the ACL for the applicable object.
-    * @param s3AsyncClient an implicit instance of a [[S3AsyncClient]].
-    * @param objectLockEnabledForBucket specifies whether you want S3 Object Lock to be enabled for the new bucket.
-    * @return a [[Task]] with the [[CreateBucketResponse]].
+    * @see how does the expected `.conf` file should look like
+    *      https://github.com/monix/monix-connect/blob/master/aws-auth/src/main/resources/reference.conf`
+    *
+    * @see the cats effect resource data type: https://typelevel.org/cats-effect/datatypes/resource.html
+    *
+    * @return a [[Resource]] of [[Task]] that allocates and releases [[S3]].
+    */
+  def fromConfig: Resource[Task, S3] = {
+    Resource.make {
+      for {
+        clientConf  <- Task.evalOnce(AppConf.loadOrThrow)
```

Comment:
Why `evalOnce`?

Reply:
@Avasil it will perform better since it would read the config only once :)

Reply:
In the current implementation, each call to `fromConfig` creates a `Task` that is memoized but it is used only once
If there are N calls to `fromConfig` then there are N memoized tasks and each one is called 1 time so there's no gain

I think that could work if `fromConfig` is a `val` or this `Task` is defined outside this method though

Reply:
Yeah it won't improve if the user keeps calling `fromConfig` over different places, but if they call it only once and assign it to a value, when this value is reused the actual read of the config file will only happen once, like this:
```scala
val s3Resource = S3.fromConfig
s3Resource.use { ... }
s3Resource.use { ... }
s3Resource.use { ... }
```

Reply:
Ah, I'd advise against it because it's not referentially transparent.

When dealing with `Task`, it shouldn't matter if we use `val`, `def`, or `lazy val`. `Task.evalOnce` and `memoize` are exceptions, Alex went with "pragmatic" choice of `Task[A]` instead of `Task[Task[A]`. Although `Task.evalOnce` could use `@UnsafeBecauseImpure`

Reply:
Good point, there is already two `@Unsafe` implementations, so prefer to sort out referential transparency here :)